File: 백준/Silver/15649. N과 M （1）/N과 M （1）.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs')
  .readFileSync(filePath)
  .toString()
  .split('\n')
  .map((line) => line.trim());

const [N, M] = input[0].split(' ').map((value) => parseInt(value));

const visited = Array(N + 1).fill(false);

function backtrack(result) {
  if (result.length === M) {
    console.log(result.join(' '));
    return;
  }

  for (let i = 1; i <= N; i++) {
    if (!visited[i]) {
      visited[i] = true;
      result.push(i);
      backtrack(result);
      result.pop();
      visited[i] = false;
    }
  }
}

backtrack([]);

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split(' '))
visited = [False for _ in range(N + 1)]
result = []

def dfs():
    if len(result) == M:
        print(*result)
        return

    for i in range(1, N + 1):
        if not visited[i]:
            visited[i] = True
            result.append(i)

            dfs()

            visited[i] = False
            result.pop()

dfs()